scheduler/core/components/optimizer/__init__.py

```diff
@@ -20,20 +16,27 @@ class Optimizer:
     """
 
     def __init__(self, algorithm=None):
+        """
+        All sites schedule the same number of nights.
+        The number of nights scheduled at one time is determined by the Selection.night_indices, passed
+        to the schedule method.
+        """
         self.algorithm = algorithm
         self.selection = None
         self.period = None
         self.night_events = None
 
     def schedule(self, selection: Selection) -> List[Plans]:
-        # Create set of plans for the amount of nights
+        """
+        The night_indices are guaranteed to be a contiguous, sorted set by Selector.select.
+        If they are not, this method will cause problems.
+        """
         self.selection = selection
         self.algorithm.setup(selection)
         self.night_events = selection.night_events
-        # TODO: Assumes that all sites schedule the same amount of nights
-        # if num_nights_optimize is None:
-        self.period = len(list(self.night_events.values())[0].time_grid)
-        nights = [Plans(self.night_events, night) for night in range(self.period)]
+
+        # Create set of plans for the amount of nights
+        nights = [Plans(self.night_events, night_idx) for night_idx in self.selection.night_indices]
```

Comment:
Right now, I'm not convinced that Bryan's code is iterating over nights except for using visibility calculations.
In the original code, he sets:
```
optimizer.periods = 3
```
but then when we call `optimizer.schedule`, it resets the `periods` to the length of the `time_grid`.

The initial code returns a set of three plans, but the second and third are always empty, so while GM is using `periods` nights for visibility calculations, it isn't (necessarily) actually calculating plans for `periods` nights.

Instead of doing that, what I do now is specify the `NightIndex` values for which we want to get a `Selection`. Since I haven't really changed any fundamental logic in `GreedyMax`, it is only using this logic for visibility calculations and calculating the plan for the first `NightIndex` in the `Selection`.

For  now, we need to focus on getting this out, so the way I see it, in `run_greedymax.py`, we iterate over the `NightIndex` subsets. Ultimately, what we will do is this, but using a loop over the `Selector.num_nights_to_schedule`:

```
plans = {}

selection = selector.select(night_indices=np.array([0, 1, 2]))
plans[0] = optimizer.schedule(selection)[0]
time accounting

selection = selector.select(night_indices=np.array([1, 2]))
plans[0] = optimizer.schedule(selection)[1]
time accounting

selection = selector.select(night_indices=np.array([2]))
plans[0] = optimizer.schedule(selection)[2]
time accounting
```

and then we will have plans for all nights across both sites.

To get GreedyMax to schedule plans for multiple nights in one call to `optimize` isn't the highest priority right now, and while I haven't had time to go over the logic in detail, since everything is scheduled on the first night and nothing is scheduled on nights 2 and 3, either in the original code or in the modifications, I think this is something we will have to look into later.